### PR TITLE
chore: Added FF to facilitate Modelgen Migration

### PR DESCRIPTION
**Pull request by akshbhu**
Description of changes:

Added FF for Codegen Migration

Added test to check for migrating package

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-codegen/package.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
     "test-watch": "jest --watch"
   },
   "dependencies": {
+    "@aws-amplify/appsync-modelgen-plugin": "^1.20.12-dev.136",
```
Comment (AaronZyLee) [inline packages/amplify-codegen/package.json:23]: The version number should be "@latest"

File: packages/amplify-codegen/src/utils/getCodegenPackageName.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const CODEGEN_OLD_PACKAGE = 'amplify-codegen-appsync-model-plugin';
+const CODEGEN_NEW_PACKAGE = '@aws-amplify/app-sync-model-gen-plugin';
```
Comment (AaronZyLee) [inline packages/amplify-codegen/src/utils/getCodegenPackageName.js:2]: Typo: new package should be '@aws-amplify/appsync-modelgen-plugin'

File: yarn.lock [hunk 3]
```diff
@@ -87,6 +87,23 @@
     "@aws-amplify/api-graphql" "^1.2.5"
     "@aws-amplify/api-rest" "^1.2.5"
 
+"@aws-amplify/appsync-modelgen-plugin@^1.20.12-dev.136":
```
Comment (AaronZyLee) [inline yarn.lock:90]: The same for version change here. It should be the latest version

**Review by AaronZyLee — CHANGES_REQUESTED** [review]

File: packages/amplify-codegen/package.json (hunk 1 above)
Comment (akshbhu) [reply packages/amplify-codegen/package.json]: changes to latest

File: yarn.lock (hunk 3 above)
Comment (akshbhu) [reply yarn.lock]: done!!

File: packages/amplify-codegen/src/utils/getCodegenPackageName.js (hunk 2 above)
Comment (akshbhu) [reply packages/amplify-codegen/src/utils/getCodegenPackageName.js]: done

File: packages/amplify-codegen/package.json (hunk 1 above)
Comment (phani-srikar) [reply packages/amplify-codegen/package.json]: I was reading about this and it is considered a good practice to not always depend on latest version of dependency package but lock it down to the latest major version. If we want to do that, we have to change it to `"@aws-amplify/appsync-modelgen-plugin": "^1.x"`. @yuth what do you think?

**Review by phani-srikar — CHANGES_REQUESTED** [review]
Some clarification on specifying dependency. Otherwise looks good.

File: packages/amplify-codegen/package.json [hunk 4]
```diff
@@ -20,6 +20,7 @@
     "test-watch": "jest --watch"
   },
   "dependencies": {
+    "@aws-amplify/appsync-modelgen-plugin": "latest",
```
Comment (phani-srikar) [inline packages/amplify-codegen/package.json:23]: I was reading about this and it is considered a good practice to not always depend on latest version of dependency package but lock it down to the latest major version. If we want to do that, we have to change it to "@aws-amplify/appsync-modelgen-plugin": "^1.x". @yuth what do you think?

File: yarn.lock [hunk 5]
```diff
@@ -6092,28 +6115,6 @@ amdefine@>=0.0.4:
   resolved "https://registry.yarnpkg.com/amdefine/-/amdefine-1.0.1.tgz#4a5282ac164729e93619bcfd3ad151f817ce91f5"
   integrity sha1-SlKCrBZHKek2Gbz9OtFR+BfOkfU=
 
-amplify-cli-core@1.3.0:
```
Comment (AaronZyLee) [inline yarn.lock:6095]: Are these changes related? Not sure this one should be removed

File: packages/amplify-codegen/src/utils/getCodegenPackageName.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+const CODEGEN_OLD_PACKAGE = 'amplify-codegen-appsync-model-plugin';
+const CODEGEN_NEW_PACKAGE = '@aws-amplify/appsync-modelgen-plugin';
+
+function getCodegenPackageName(isNewCodegenPackage) {
```
Comment (yuth) [inline packages/amplify-codegen/src/utils/getCodegenPackageName.js:4]: I think it would be better to require both the module and then send the right one based on the flag. If the packages are missing dynamic imports defer the problem till the specific code path is hit

**Review by yuth — CHANGES_REQUESTED** [review]

File: packages/amplify-codegen/package.json (hunk 4 above)
Comment (akshbhu) [reply packages/amplify-codegen/package.json]: done

File: packages/amplify-cli-core/src/feature-flags/featureFlags.ts [hunk 7]
```diff
@@ -511,5 +511,16 @@ export class FeatureFlags {
         defaultValueForNewProjects: false,
       },
     ]);
+
+    this.registerFlag('codegenPackageMigration', [
```
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-cli-core/src/feature-flags/featureFlags.ts]: done

File: packages/amplify-codegen/src/utils/getCodegenPackageName.js (hunk 6 above)
Comment (akshbhu) [reply packages/amplify-codegen/src/utils/getCodegenPackageName.js:4]: done

Comment (yuth) [reply packages/amplify-codegen/src/utils/getCodegenPackageName.js:4]: Can you import this at  the top level. Otherwise there is no benefit doing this

Comment (akshbhu) [reply packages/amplify-codegen/src/utils/getCodegenPackageName.js]: import moved to top and added a custom property "name" in the mockModule

File: packages/amplify-codegen/tests/utils/getCodegenPackage.test.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+// old codegen package
+jest.mock('amplify-codegen-appsync-model-plugin', () => ({
+  __esModule: true,
+  addToSchema: jest.fn(),
+  plugin: jest.fn(),
+  preset: {
+    buildGeneratesSection: jest.fn(),
+  },
+  name: 'OldCodegenPackage',
+}));
```
Comment (yuth) [inline packages/amplify-codegen/tests/utils/getCodegenPackage.test.js:35]:
```suggestion
jest.mock('amplify-codegen-appsync-model-plugin', () => oldCodegenPackage_mock);
```

File: packages/amplify-codegen/tests/utils/getCodegenPackage.test.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+// new codegen package
+jest.mock('@aws-amplify/appsync-modelgen-plugin', () => ({
+  __esModule: true,
+  addToSchema: jest.fn(),
+  plugin: jest.fn(),
+  preset: {
+    buildGeneratesSection: jest.fn(),
+  },
+  name: 'NewCodegenPackage',
+}));
```
Comment (yuth) [inline packages/amplify-codegen/tests/utils/getCodegenPackage.test.js:46]:
```suggestion
jest.mock('@aws-amplify/appsync-modelgen-plugin', () => newCodegenPackage_mock);
```

**Review by yuth — CHANGES_REQUESTED** [review]

**Review by yuth — APPROVED** [review]

**Review by AaronZyLee — APPROVED** [review]

**Review by phani-srikar — APPROVED** [review]
Thanks for these changes @akshbhu :)
